### PR TITLE
Improve test coverage of pkg/image

### PR DESCRIPTION
**Pull request by gvallee**
**Description of the Pull Request (PR):**

Improve test coverage of pkg/image.

**This fixes or addresses the following GitHub issues:**

- Addresses #2559 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by gvallee** [comment]
Addresses #3309

File: pkg/image/ext3_test.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	"testing"
+)
+
+// createVirtualBlockDevice creates a virtual block device
+// in a file using the dd command. The test is skipped if
+// the command is not available.
+// @parma[in] test handle to control the test, i.e., stop it in case of fatal error
+// @parma[in] path to the virtual block device to be created
+func createVirtualBlockDevice(t *testing.T, path string) {
+	cmdBin := "/bin/dd"
```
Comment (cclerget) [inline pkg/image/ext3_test.go]: You can use `exec.LookPath` instead of an hardcoded path then you can remove `os.Stat` check

File: pkg/image/ext3_test.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+// Create a new file system in an existing virtual block
+// device. This function relies on mkfs.X to create the
+// file system, the test is skipping if the command is
+// not available.
+// @param[in] test handle to control the test, i.e., stop it in case of fatal error
+// @param[in] type of the file system to be created, e.g., ext3, fat.
+// @param[in] path to the virtual block device.
+func createFS(t *testing.T, fsType string, path string) {
+	cmdBin := "/sbin/mke2fs"
```
Comment (cclerget) [inline pkg/image/ext3_test.go]: same here with `exec.LookPath`

File: pkg/image/ext3_test.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	_, checkErr := CheckExt3Header(b)
+	if checkErr != nil {
+		t.Fatalf("cannot check ext3 header of a valid image: %s\n", checkErr)
+	}
+}
+
+func TestInitializer(t *testing.T) {
+	// Create a temporary image which is obviously an invalid ext3 image
```
Comment (cclerget) [inline pkg/image/ext3_test.go:140]: Tests not requiring privileges must drop privileges at the beginning of the function with snippet :

```
test.DropPrivilege(t)
defer test.ResetPrivilege(t)
```

Otherwise they could be executed as root, because `go test` is called with sudo, so all tests function must explicitly tell if they need to keep privileges or drop privileges during execution

For functions requiring privileges, a call `test.EnsurePrivilege(t)` at the beginning of the function is required

File: pkg/image/ext3_test.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+	f, err := ioutil.TempFile("", "image-")
+	if err != nil {
+		t.Fatalf("cannot create temporary file: %s\n", err)
+	}
+	path, err := filepath.Abs(f.Name())
```
Comment (cclerget) [inline pkg/image/ext3_test.go]: `f.Name()` already returns the absolute path

File: pkg/image/ext3_test.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+	// We now run a test with ext2 to hit some other corner cases
+	err = ext3InitializerTest(t, img, resolvedPath, "ext2")
+	if err == nil {
+		t.Fatal("ext3 initializer test succeeded with an ext2 image while expected to fail")
+	}
+
+	// We reformat the image with different file systems and see if we catch
+	// the error
+	_, statErr := os.Stat("/sbin/mkfs.fat")
```
Comment (cclerget) [inline pkg/image/ext3_test.go]: `exec.LookPath("mkfs.fat")`

File: pkg/image/ext3_test.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+	if statErr == nil {
+		err = ext3InitializerTest(t, img, resolvedPath, "vfat")
+		if err == nil {
+			t.Fatal("ext3 initializer test succeeded with a vfat image")
+		}
+	} else {
+		t.Log("/sbin/mkfs.fat is not available, skipping the test...")
+	}
+
+	_, statErr = os.Stat("/sbin/mkfs.ext4")
```
Comment (cclerget) [inline pkg/image/ext3_test.go]: `exec.LookPath("mkfs.ext4")`

File: pkg/image/image_test.go [hunk 7]
```diff
@@ -150,3 +152,154 @@ func TestReader(t *testing.T) {
...
+
+func TestAuthorizedOwner(t *testing.T) {
+	type ownerGroup struct {
+		name       string
+		owners     []string
+		shouldPass bool
+	}
+
+	// Note that we do not test the "root" case because the result will depend
+	// on the context of the test execution and therefore sometimes pass,
```
Comment (cclerget) [inline pkg/image/image_test.go]: Calling `test.EnsurePrivilege(t)` at the beginning of this function should fix that

File: pkg/image/image_test.go [hunk 8]
```diff
@@ -150,3 +152,154 @@ func TestReader(t *testing.T) {
...
+
+func TestAuthorizedGroup(t *testing.T) {
+	type groupTest struct {
+		name       string
+		groups     []string
+		shouldPass bool
+	}
+
+	// Note that we do not test the "root" case because the result will depend
+	// on the context of the test execution and therefore sometimes pass,
```
Comment (cclerget) [inline pkg/image/image_test.go]: same comment

File: pkg/image/sandbox_test.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	// Invalid case using a file
+	f, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatalf("cannot create temporary file: %s\n", err)
+	}
+	defer f.Close()
+
+	path, err = filepath.Abs(f.Name())
```
Comment (cclerget) [inline pkg/image/sandbox_test.go]: not required

File: pkg/image/squashfs_test.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	defer os.RemoveAll(dir)
+
+	sqshFile, fileErr := ioutil.TempFile("", "")
+	if fileErr != nil {
+		t.Fatalf("impossible to create temporary file: %s\n", fileErr)
+	}
+	defer sqshFile.Close()
+
+	sqshFilePath, err := filepath.Abs(sqshFile.Name())
```
Comment (cclerget) [inline pkg/image/squashfs_test.go]: not required

File: pkg/image/squashfs_test.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		t.Fatalf("cannot get path of temporary file: %s\n", err)
+	}
+	// close and delete the temp file since we will be used with the
+	// mksquashfs command. We still use TempFile to have a clean way
+	// to get a valid path for a temporary file
+	sqshFile.Close()
+	os.Remove(sqshFilePath)
+
+	cmdBin := "/usr/bin/mksquashfs"
```
Comment (cclerget) [inline pkg/image/squashfs_test.go]: `exec.LookPath("mksquashfs")`

**Review by cclerget — CHANGES_REQUESTED** [review]

File: pkg/image/ext3_test.go (hunk 3 above)
Comment (gvallee) [reply pkg/image/ext3_test.go:140]: Good to know! Should we audit the tests and make sure we actually do that? I do not recall seeing this since I started working on testing. What do you think?

File: pkg/image/ext3_test.go (hunk 1 above)
Comment (gvallee) [reply pkg/image/ext3_test.go]: fixed

File: pkg/image/ext3_test.go (hunk 2 above)
Comment (gvallee) [reply pkg/image/ext3_test.go]: fixed

File: pkg/image/ext3_test.go (hunk 3 above)
Comment (gvallee) [reply pkg/image/ext3_test.go:140]: I revisit the entire code to rely on this. These tests should now be good to go.

File: pkg/image/ext3_test.go (hunk 4 above)
Comment (gvallee) [reply pkg/image/ext3_test.go]: fixed

File: pkg/image/ext3_test.go (hunk 5 above)
Comment (gvallee) [reply pkg/image/ext3_test.go]: fixed

File: pkg/image/ext3_test.go (hunk 6 above)
Comment (gvallee) [reply pkg/image/ext3_test.go]: fixed

File: pkg/image/image_test.go (hunk 7 above)
Comment (gvallee) [reply pkg/image/image_test.go]: fixed

File: pkg/image/image_test.go (hunk 8 above)
Comment (gvallee) [reply pkg/image/image_test.go]: fixed

File: pkg/image/sandbox_test.go (hunk 9 above)
Comment (gvallee) [reply pkg/image/sandbox_test.go]: fixed

File: pkg/image/squashfs_test.go (hunk 10 above)
Comment (gvallee) [reply pkg/image/squashfs_test.go]: fixed

File: pkg/image/squashfs_test.go (hunk 11 above)
Comment (gvallee) [reply pkg/image/squashfs_test.go]: fixed

File: pkg/image/ext3_test.go (hunk 3 above)
Comment (cclerget) [reply pkg/image/ext3_test.go:140]: To be consistent yes, will open an issue and add it to epic

**Review by cclerget — APPROVED** [review]
